### PR TITLE
FIX: fix interpolate with kwarg limit area and limit direction using pad or bfill

### LINKED ISSUE #26796: limit_area and limit_direction do not have an effect when interpolation method is 'pad'
#### Code Sample, a copy-pastable example if possible

```python
In [1]: import pandas as pd                                                                                                                                                                                                        

In [2]: import numpy as np                                                                                                                                                                                                         

In [3]: s = pd.Series([ 
   ...:     np.nan, 
   ...:     1., np.nan, 
   ...:     2., np.nan, np.nan, 
   ...:     5., np.nan, np.nan, np.nan, 
   ...:     -1., np.nan, np.nan 
   ...: ])                                                                                                                                                                                                                         

In [4]: s.interpolate(method='pad', limit_area='inside')                                                                                                                                                                           
Out[4]: 
0     NaN
1     1.0
2     1.0
3     2.0
4     2.0
5     2.0
6     5.0
7     5.0
8     5.0
9     5.0
10   -1.0
11   -1.0
12   -1.0
dtype: float64

# For method='linear' the `limit_area` kwarg works as expected

In [5]: s.interpolate(method='linear', limit_area='inside')                                                                                                                                                                        
Out[5]: 
0     NaN
1     1.0
2     1.5
3     2.0
4     3.0
5     4.0
6     5.0
7     3.5
8     2.0
9     0.5
10   -1.0
11    NaN
12    NaN
dtype: float64


```
#### Problem description

The kwargs `limit_area` and `limit_direction` for `interpolate()`, introduce in https://github.com/pandas-dev/pandas/pull/16513 do not seem to have an effect when using the method `pad`. They work with other interpolation methods, e.g. `linear`.

There are two different pathways for interpolate depending on the selected method. 
https://github.com/pandas-dev/pandas/blob/b8ad9da4213be55c587db015c640d927aeb7feaf/pandas/core/internals/blocks.py#L1096-L1116

For pad, ffill and bfill `_interpolate_with_fill()` is used which calls `missing.interpolate_2d` which does not seem to recognize the keywords limit_direction and limit_area. They are silently ignored.

I might be able to fix that during https://github.com/pandas-dev/pandas/pull/25141, but maybe a fix needs more fundamental changes.

#### Expected Output

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: eaacefd0957bfc6735c8782ac2ba571d7131d250
python: 3.7.2.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: None
LOCALE: de_DE.UTF-8

pandas: 0.25.0.dev0+725.geaacefd09
pytest: 4.1.1
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.2
numpy: 1.15.4
scipy: 1.2.0
pyarrow: 0.11.1
xarray: 0.11.0
IPython: 7.2.0
sphinx: 1.8.2
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.9
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.12
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml.etree: 4.3.0
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: 1.2.16
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: 0.2.0
fastparquet: 0.2.1
pandas_gbq: None
pandas_datareader: None
gcsfs: None


</details>

### PR DESCRIPTION
**Pull request by cchwala**
- [x] closes #26796
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This separates a large part of the code changes from #25141 to keep things comprehendible.

### CODE REVIEW

**Comment by WillAyd** [comment]
@cchwala so IIUC you are doing a precursor to solve the issue with pad and `limit_area` and then planning to add `max_gap` after that gets merged in the original PR right?

**Comment by cchwala** [comment]
@WillAyd Yes. That is my plan. 

I am reusing the code from the PR with `max_gap` that actually solves #26796. The `max_gap` PR will be a lot smaller when the changes from this PR are merge.

File: pandas/core/missing.py
Comment (WillAyd) [inline pandas/core/missing.py]:
```suggestion
    method: str = "pad",
```

File: pandas/core/missing.py
Comment (WillAyd) [inline pandas/core/missing.py]:
```suggestion
    axis: Axis = 0,
```

(import from pandas._typing)

File: pandas/core/missing.py
Comment (WillAyd) [inline pandas/core/missing.py]:
```suggestion
    limit: Optional[int] = None,
```

File: pandas/core/missing.py
Comment (WillAyd) [inline pandas/core/missing.py]:
```suggestion
    limit_area: Optional[str] = None,
```

File: pandas/core/missing.py
Comment (WillAyd) [inline pandas/core/missing.py]:
```suggestion
    dtype: Optional[Dtype] = None,
```

(import from pandas._typing)

File: pandas/core/missing.py
Comment (WillAyd) [inline pandas/core/missing.py]:
```suggestion
    fill_value: Optional[Hashable] = None,
```

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks pretty good. If you can annotate any new functions would be helpful. Added what I think should be

**Comment by cchwala** [comment]
Should be ready now. I am offline for some hours of sleep, though.

**Comment by cchwala** [comment]
@WillAyd Any further request for changes?

File: pandas/core/generic.py
Comment (WillAyd) [inline pandas/core/generic.py]: Should revert this; though it shows `None` as the value it still does default to `forward` no?

File: pandas/core/missing.py
Comment (jreback) [inline pandas/core/missing.py]: can you try to type these * add a Paramaters / Returns section

File: pandas/core/internals/blocks.py
Comment (jreback) [inline pandas/core/internals/blocks.py]: instead of handling this case here and distributing the logic, can you simply call it from interpolate_2d?

File: pandas/core/generic.py
Comment (jreback) [inline pandas/core/generic.py]: is there a reason this logic does not simply belong in interpolate_2d?

Comment (jreback) [reply pandas/core/generic.py]: or call something like clean_fill_method

File: pandas/core/missing.py
Comment (jreback) [inline pandas/core/missing.py]: can you move this under interpolate_1d

File: doc/source/whatsnew/v1.0.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: move this to 1.1

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good

**Comment by WillAyd** [comment]
@cchwala can you address comments?

File: pandas/core/generic.py
Comment (cchwala) [reply pandas/core/generic.py]: Thanks for spotting this. Will be fixed in next series of commits.

File: pandas/core/missing.py
Comment (cchwala) [reply pandas/core/missing.py]: Yes. Will be update in next series of commits

File: pandas/core/internals/blocks.py
Comment (cchwala) [reply pandas/core/internals/blocks.py]: Yes. Will be moved in next series of commits.

File: doc/source/whatsnew/v1.0.0.rst
Comment (cchwala) [reply doc/source/whatsnew/v1.0.0.rst]: done

**Comment by cchwala** [comment]
@WillAyd Sorry for the delay. I adjusted the code according to most comments. I am now looking into the last two open comments and will respond/commit when I have an answer.

File: pandas/core/missing.py
Comment (cchwala) [reply pandas/core/missing.py]: `interpolate_1d_fill` is a special case needed by `blocks._interpolate_with_fill()` to support the limit kwargs with method `pad`. As you requested in the comment above `interpolate_1d_fill` is now called from within `interpolate_2d` which is what `blocks._interpolate_with_fill()` calls.

You request, moving it to `missing.interpolate_1d` will, in my opinion, not work, because `missing.interpolate_1d` is not reached in the call sequence for interpolating with method `pad` since `blocks.Block.interpolate()` splits up into `blocks.Block._interpolate` (for scipy wrappers) and `blocks.Block._interpolate_with_fill` (for pad methods). `missing.interpolate_1d` is only called in the former case, i.e. for methods using the scipy wrappers.

File: pandas/core/generic.py
Comment (cchwala) [reply pandas/core/generic.py]: `limit_direction` has no effect in `interpolate_2d`.

My idea was to detect conflicting user input, e.g. `pad` with `limit_direction='backward'` as early as possible in the chain of calls to the main functions that are involved.

But I can move the logic to a function like `clean_fill_method`, e.g. called `set_limit_direction`. Should this function belong to generic.py or missing.py?

File: pandas/core/missing.py
Comment (cchwala) [inline pandas/core/missing.py]: @jreback Should I force a set here with
```
preserve_nans = set(sorted(preserve_nans))
```
so that the CI type validation does not complain?

**Comment by cchwala** [comment]
@WillAyd @jreback Can you comment on the changes. I would potentially have time to further work on this and the related https://github.com/pandas-dev/pandas/pull/25141, which has to wait till this is merged.

File: pandas/core/missing.py
Comment (WillAyd) [reply pandas/core/missing.py]: If you don't re-use the same variable should be OK, so something like `nans_list = sorted(preserve_nans)` should work

Comment (cchwala) [reply pandas/core/missing.py]: Solved

**Comment by cchwala** [comment]
@WillAyd I resolved the conflicts with upstream/master so that this PR can potentially be merged.

File: pandas/core/missing.py
Comment (WillAyd) [inline pandas/core/missing.py]:
```suggestion
    preserve_nans: list of int
```

sorted converts the set to a list

File: pandas/core/missing.py
Comment (WillAyd) [inline pandas/core/missing.py]:
```suggestion
) -> List[int]:
```

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by cchwala** [comment]
@WillAyd Sorry for missing the details of the required changes. These are solved now.

File: pandas/tests/series/methods/test_interpolate.py
Comment (WillAyd) [inline pandas/tests/series/methods/test_interpolate.py]: These should really be parametrized...but not going to block since this PR has gone on for a long time (great job on keeping updated by the way)

**Review by WillAyd — APPROVED** [review]
small comment that can be address in a follow on but otherwise lgtm. @jreback

**Comment by jreback** [comment]
@cchwala can you merge master.

**Comment by simonjayhawkins** [comment]
@cchwala can you merge master

File: doc/source/whatsnew/v1.1.0.rst
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.1.0.rst:311]: issue #26796

File: pandas/core/generic.py
Comment (simonjayhawkins) [inline pandas/core/generic.py:6674]: the default is now 'backward' for method 'backfill' and 'bfill'?

**Comment by jreback** [comment]
@simonjayhawkins I think this is worth merge *after* your set cleaning in interpolate.

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins I think this is worth merge _after_ your set cleaning in interpolate.

i've done this locally but get
```
$ git push
fatal: The upstream branch of your current branch does not match
the name of your current branch.  To push to the upstream branch
on the remote, use

    git push upstream HEAD:refs/pull/31048/head

To push to the branch of the same name on the remote, use

    git push upstream fix_interpolate_limit_area_and_limit_direction_with_pad

To choose either option permanently, see push.default in 'git help config'.
```

am using hub.

BUT tests are failing. I think related to changes in #33959. not looked yet.

File: pandas/core/generic.py
Comment (simonjayhawkins) [reply pandas/core/generic.py:6869]: interpolate_2d is also called from fillna. only NDFrame.interpolate allows limit_direction so does make sense to validate here.

clean_fill_method checks method. probably want to avoid passing more parameters around.

File: pandas/core/missing.py
Comment (simonjayhawkins) [reply pandas/core/missing.py:521]: @cchwala interpolate_2d will work on a 1d array. did you investigate applying it along an axis (with masking logic) rather than creating a 1d version.

File: pandas/tests/series/methods/test_interpolate.py
Comment (simonjayhawkins) [reply pandas/tests/series/methods/test_interpolate.py:453]: done in #34746

**Comment by simonjayhawkins** [comment]
> BUT tests are failing. I think related to changes in #33959. not looked yet.

will need to make some changes to the implementation of #33959, will raise a PR shortly based on changes in #34749

File: pandas/core/missing.py
Comment (simonjayhawkins) [reply pandas/core/missing.py:521]: see #34749 for alternative implementation calling interpolate_2d instead. interpolate_2d already has the limit logic so no need to use the preserve_nans set based logic.

**Comment by cchwala** [comment]
@simonjayhawkins Sorry for not responding. I was, and still am, very busy with other things. I am available to help with this PR from July on.

File: doc/source/whatsnew/v1.1.0.rst
Comment (jreback) [reply doc/source/whatsnew/v1.1.0.rst:311]: pls move to 1.2

File: pandas/core/generic.py
Comment (jreback) [reply pandas/core/generic.py:6674]: @cchwala can you address

File: pandas/core/generic.py
Comment (jreback) [inline pandas/core/generic.py:6829]: can you type this (other keywords are a +)

File: pandas/core/generic.py
Comment (jreback) [reply pandas/core/generic.py:6869]: pls make a method similr to clean_fill_method then, e.g this logic should living pandas/core/missing.py

File: pandas/core/missing.py
Comment (jreback) [inline pandas/core/missing.py:321]: better to use numpy set ops here, no?

e.g. np.unique

**Review by jreback — CHANGES_REQUESTED** [review]
not really sure of the status here and @simonjayhawkins had some cleanups as well, can we clarify.

**Comment by simonjayhawkins** [comment]
> not really sure of the status here and @simonjayhawkins had some cleanups as well, can we clarify.

I did break some bits off this PR. so does need a merge of master to see what's left (I did try this https://github.com/pandas-dev/pandas/pull/31048#issuecomment-643394779)

**Comment by cchwala** [comment]
Okay. I will do the merge with master today to see what is left. I guess, at least the tests will still be of value.

**Comment by cchwala** [comment]
I am now looking into [the suggestion](https://github.com/pandas-dev/pandas/pull/31048/files/e259549e2f369bd07a731681d1278e1e5d605007#r439754395) from @simonjayhawkins to use `interpolate_2d` instead of introducing `interpolate_1d_fill`

File: pandas/core/generic.py
Comment (cchwala) [reply pandas/core/generic.py]: not relevant anymore

File: pandas/core/generic.py
Comment (cchwala) [reply pandas/core/generic.py]: not sure what to do here. after merge with master, there is similar code at this location, but it is not mine...

File: pandas/core/missing.py
Comment (cchwala) [reply pandas/core/missing.py]: not relevant anymore

File: pandas/core/missing.py
Comment (cchwala) [reply pandas/core/missing.py]: I used the nice solution from #34749 as suggested above

Comment (cchwala) [reply pandas/core/missing.py]: as a consequence there is no `interpolate_1d_fill` anymore

File: pandas/core/generic.py
Comment (cchwala) [reply pandas/core/generic.py]: already change after merging master

**Comment by cchwala** [comment]
@jreback can you please have another look? As expected there is not much left from this PR because parts of it have already been merge elsewhere. Most of your code review comments do not apply anymore, only [this one](https://github.com/pandas-dev/pandas/pull/31048/files/e259549e2f369bd07a731681d1278e1e5d605007#r477598954).

**Comment by cchwala** [comment]
@jreback Sorry to ping you again, but it would be great if I could finish this PR with (hopefully) final adjustments tomorrow (Friday). Otherwise things would have to wait again another two weeks because I will be offline.

**Comment by jreback** [comment]
will have to wait as i am away

**Comment by cchwala** [comment]
Okay. No problem.

File: pandas/core/internals/blocks.py
Comment (jreback) [inline pandas/core/internals/blocks.py:1190]: can you put this in a function in `missing` and just call here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py
Comment (cchwala) [reply pandas/core/internals/blocks.py:1190]: Okay. But if I do this I could also directly integrate it into `missing.interpolate_2d`, e.g. like [in this branch](https://github.com/pandas-dev/pandas/compare/master...cchwala:quick_test_interpolate_limit_area). Advantage, no need to introduce yet another function that does some interpolation. Disadvantage, obfuscates `missing.interpolate_2d`. For this, adding some comments would help for sure.

If you prefer the separate function in `missing`, what function name would you choose?

Comment (simonjayhawkins) [reply pandas/core/internals/blocks.py:1190]: > e.g. like [in this branch](https://github.com/pandas-dev/pandas/compare/master...cchwala:quick_test_interpolate_limit_area).

I would merge recursive solution here for review (so that ci has been run against it, i think needs changes to pandas/core/arrays/categorical.py see https://github.com/pandas-dev/pandas/compare/master...simonjayhawkins:fix_interpolate_limit_area_and_limit_direction_with_pad)

Comment (cchwala) [reply pandas/core/internals/blocks.py]: Okay. Merged and CI is running.

I have also pushed [a branch](https://github.com/pandas-dev/pandas/compare/master...cchwala:quick_test_limit_area_option2) with an example implementation using a separate function in `missing`. The name of the new function is not good, but the implementation would probably look something like this.

For both solutions, tests for interpolation pass, but I did not run the full test suite. Will see what changes are required when CI is done.

Comment (cchwala) [reply pandas/core/internals/blocks.py]: Test are green. 

If you find the recursive solution appropriate, I would add a comment to the function that briefly explains the reasoning. Anything else that needs improvement?

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @cchwala can you add the type annotations as requested by @WillAyd https://github.com/pandas-dev/pandas/pull/31048/files#r367123208

File: pandas/core/internals/blocks.py
Comment (simonjayhawkins) [reply pandas/core/internals/blocks.py]: @jreback

**Comment by simonjayhawkins** [comment]
@cchwala can you resolve merge conflicts and can take another look

File: pandas/core/missing.py
Comment (jreback) [inline pandas/core/missing.py:565]: can you make this a module level function and just call it here

**Review by jreback — CHANGES_REQUESTED** [review]
looks ok, some comments, can you merge master and we can get this in.

**Comment by jreback** [comment]
moving off 1.2 need merge master and update to comments.

**Comment by arw2019** [comment]
Closing in favor of #38106

**Pull request closed without merging**
